clojure/src/ch02_lisp2.clj

```diff
@@ -0,0 +1,182 @@
+
+
+;;; In Lisp-2, we will restrict the first element of every form to be a symbol.
+;;; That simplifies  the evaluation for the term in function position a lot.
+;;; We no longer needs all the complexity of `evaluate`.
+
+;; Let's copy `evaluate` but modify the function invocation at the end
+;; - call `evaluate-application` instead of using `invoke` directly
+(declare evaluate-application f-eprogn f-evlis f-make-function)
+(defn f-evaluate [exp env fenv]
+  (if (e/atom? exp)
+    (cond
+      ;; lock immutability of t and f in the interpreter
+      (= 't exp) true
+      (= 'f exp) false
+      (symbol? exp) (e/lookup exp env)
+      ;; Notice that `keyword?` isn't here because keywords are Clojure's thing
+      ;; and aren't present in the Lisp we are trying to implement
+      ((some-fn number? string? char? boolean? vector?) exp) exp
+      :else (e/wrong "Cannot evaluate - unknown atomic expression?" exp))
+
+    ;; we use `first` instead of `car`
+    (case (first exp)
+      quote (second exp)
+      ;; (p.8) we gloss over the fact that in `(if pred)` we use boolean semantics
+      ;; of the implementation language (Clojure - which means `nil` will be falsy);
+      ;; more precisely, we should write `(if-not (= the-false-value (evaluate (second exp) env)))
+      if (if (f-evaluate (second exp) env fenv)
+           (f-evaluate (nth exp 2) env fenv)
+           (f-evaluate (nth exp 3) env fenv))
+      begin (f-eprogn (rest exp) env fenv)
+      set! (e/update! (second exp) env (f-evaluate (nth exp 2) env fenv))
+      lambda (f-make-function (second exp) (nnext exp) env fenv)
+      ;; CHANGE: call `evaluate-application` instead of `invoke`
+      (evaluate-application (first exp)
+                            (f-evlis (rest exp) env fenv)
+                            env
+                            fenv))))
+
+;; ... we need to re-define the helper functions like f-eprogn and f-evlis
+;; to make sure their propagate the function environment `fenv`
+(defn f-evlis [exps env fenv]
+  (if (list? exps)
+    (map #(f-evaluate % env fenv) exps)
+    ()))
+
+(defn f-eprogn
+  "Evaluates sequence of expressions in given environment."
+  [exps env fenv]
+  (if (list? exps)
+    (let [[fst & rst] exps
+          fst-val (f-evaluate fst env fenv)]
+      (if (list? rst)
+        (f-eprogn rst env fenv) 
+        fst-val))
+    ()))
+
+;; ... now we redefine `make-function` to leverage `fenv`
+(defn f-make-function [variables body env fenv]
+  (fn [values]
+    ;; notice that `fenv` isn't extended since it's only for functions,
+    ;; not for the variables inside the body of the function
+    (f-eprogn body (e/extend env variables values) fenv)))
+
+;; ... and finally we can implement `evaluate-application`
+(defn evaluate-application [f args env fenv]
+  (cond
+    ;; function symbol -> invoke it right away
+    (symbol? f)
+    (e/invoke (e/lookup f fenv) args) ; notice we lookup the function symbol in `fenv`
```

Comment:
Call to `e/invoke` can be eliminated because we know all the symbols in `fenv` are functions => we can just look them up and invoke immediately.